### PR TITLE
INT-2367 - Add Commit Info

### PR DESCRIPTION
**Pull request by adam-in-ict**
Adding MR commit info.  

At this time, there is no durable way to create and maintain relationships between commits and users, so the author and committer data currently resides only on the commit entity itself.  In the future if the Gitlab API changes to provide a user ID for commit data, we can modify this to create the correct relationships between commits and users.

### CODE REVIEW

File: src/steps/commits.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+  const errorCorrelationId = uuid();
+
+  await jobState.iterateEntities(
+    { _type: Entities.MERGE_REQUEST._type },
+    async (mergeRequestEntity) => {
+      const mergeRequest = getRawData(mergeRequestEntity) as GitLabMergeRequest;
+      await client.iterateMergeRequestCommits(
+        mergeRequest.project_id,
+        mergeRequest.iid,
```
Comment (ndowmon) [inline src/steps/commits.ts:32]: Is this `mergeRequest.iid` a typo? 

I haven't been in this project much but I'm surprised to see that we don't have typing for:
```ts
export type GitLabProject = Opaque<any, 'GitLabProject'>;
export type GitLabMergeRequest = Opaque<any, 'GitLabMergeRequest'>;
export type GitLabMergeCommitRequest = Opaque<any, 'GitLabMergeCommitRequest'>;
```

I'd love to see actual typing for `GitLabMergeCommitRequest`,  even if all of the properties are optional.

File: src/steps/commits.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                logDetail,
+                'Unauthorized to fetch merge request commits for project',
+              );
+            } else {
+              logger.error(
+                logDetail,
+                'Failed to fetch merge request commits for project',
+              );
+            }
+          },
```
Comment (ndowmon) [inline src/steps/commits.ts:64]: Is this `onPageError` arg a standard pattern in this project? Seems like the kind of logic that should be encapsulated within the client itself, rather than provided by the consumer?

File: src/steps/commits.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  );
+}
+
+export const commitSteps: IntegrationStep<GitlabIntegrationConfig>[] = [
+  {
+    id: Steps.COMMITS,
+    name: 'Fetch MR commits',
+    entities: [Entities.COMMIT],
+    relationships: [Relationships.MR_HAS_COMMIT],
+    dependsOn: [Steps.MERGE_REQUESTS, Steps.USERS],
```
Comment (ndowmon) [inline src/steps/commits.ts:77]: Not sure if `Steps.USERS` is a typo - I may be missing something in the execution handler.

Comment (ndowmon) [reply src/steps/commits.ts:77]: BTW, I was wondering if it's possible to match _signed_ commits with GitHub users? It makes perfect sense that unsigned commits cannot be correlated with a user, but it would be really interesting to see if signed commits have additional data that could enable us to build a relationship.

Screen Shot 2022-02-17 at 2 14 53 PM

File: CHANGELOG.md [hunk 4]
```diff
@@ -8,6 +8,10 @@ and this project adheres to
 
 ## [Unreleased]
 
+### Added
+- Commits for MRs are now added. They are only related to MRs at this time, as
+the required data doesn't exist to have a durable relationship creation process
+between users and commits.
```
Comment (ndowmon) [inline CHANGELOG.md:14]: I wonder if a `project -HAS-> commit` relationship is worth adding as well, at some point. There are definitely commits in a project that are not related to a merge request.

File: src/converters/index.ts [hunk 5]
```diff
@@ -160,7 +161,47 @@ export function createMergeRequestEntity(
...
+      source: mergeRequestCommit,
+      assign: {
+        _key: key,
+        _type: Entities.COMMIT._type,
+        _class: Entities.COMMIT._class,
+        id: String(mergeRequestCommit.id),
+        shortId: String(mergeRequestCommit.short_id),
+        title: mergeRequestCommit.title,
+        name: mergeRequestCommit.title,
+        branch: mergeRequest.source_branch,
```
Comment (ndowmon) [inline src/converters/index.ts:181]: This seems to be the only property from the `mergeRequest`. Is it necessary to have on the `mergeRequestCommit`? It seems a little leaky that in order to create a `Commit` entity, you need to know about the merge request it's on.

File: src/steps/commits.ts (hunk 3 above)
Comment (ndowmon) [reply src/steps/commits.ts:77]: Oh, this is GitLab, not GitHub 😅 . Does GitLab have the same concepts?

File: src/steps/commits.ts (hunk 1 above)
Comment (adam-in-ict) [reply src/steps/commits.ts:32]: Nope.  `iid` is what we want.  It definitely tripped me up a time or two.

I agree on typing, but opted to match what was currently done in this integration for the time being.  Certainly up for challenge though.

File: src/steps/commits.ts (hunk 2 above)
Comment (adam-in-ict) [reply src/steps/commits.ts:64]: Yeah, this is a bit of an older integration.  I perhaps opted too much to match the current setup instead of updating everything to something closer to what we've been doing in newer integrations.

File: src/steps/commits.ts (hunk 3 above)
Comment (adam-in-ict) [reply src/steps/commits.ts:77]: Steps.USERS is an artifact of when I thought I'd be able to create relationships between commits and users.  Currently there's no user ID data on any commits in Gitlab, so I think this is just something I need to clean up until such time as we have updates to the API that support it.

File: CHANGELOG.md (hunk 4 above)
Comment (adam-in-ict) [reply CHANGELOG.md:14]: There are, but we're currently only ingesting commits associated with MRs.  Originally the request was to actually just add this data to the gitlab_merge_request entity, but there's no way to do that when MRs can have multiple commits, so I opted to break them out into their own entities.

It does bring up a good question of whether we want to restrict by MR or pull in all commits.  We're restricting what MR data we pull in based on date, so we'd also need to decide if commits not tied to MRs would be restricted by that value as well.

File: src/converters/index.ts (hunk 5 above)
Comment (adam-in-ict) [reply src/converters/index.ts:181]: Branch is a required field for the CodeCommit _class that I assigned to this entity and Gitlab commit info doesn't list branch info.  Pulling in the mergeRequest data to get the branch seemed like the least objectionable option.

File: src/steps/commits.ts (hunk 1 above)
Comment (ndowmon) [reply src/steps/commits.ts:32]: Oh, wow you're totally right! I just read these docs: https://docs.gitlab.com/ee/api/#id-vs-iid - very confusing.

It'd be interesting to know where the integrations developers stand on the typed vs `Opaque` issue. There was a time when we thought `Opaque` was a better option, but I'm nots sure that's necessarily true anymore.

File: src/steps/commits.ts (hunk 3 above)
Comment (ndowmon) [reply src/steps/commits.ts]: 👍

File: CHANGELOG.md (hunk 4 above)
Comment (ndowmon) [reply CHANGELOG.md:14]: 👍

**Review by ndowmon — APPROVED** [review]
